cpp/include/cugraph/detail/decompress_matrix_partition.cuh

```diff
@@ -180,10 +180,247 @@ void decompress_matrix_partition_to_fill_edgelist_majors(
+       input_major_start_offsets =
+         input_major_start_offsets + segment_offsets[3] - segment_offsets[0],
+       majors,
+       minors,
+       property = property
+                    ? thrust::make_optional(thrust::make_tuple(
+                        thrust::get<0>(*property) + segment_offsets[3], thrust::get<1>(*property)))
+                    : thrust::nullopt] __device__(auto idx) {
+        auto major        = input_majors[idx];
+        auto major_offset = input_major_start_offsets[idx];
+        auto major_idx    = matrix_partition.get_major_hypersparse_idx_from_major_nocheck(major);
+        if (major_idx) {
+          vertex_t const* indices{nullptr};
+          thrust::optional<weight_t const*> weights{thrust::nullopt};
+          edge_t local_degree{};
+          thrust::tie(indices, weights, local_degree) =
+            matrix_partition.get_local_edges(*major_idx);
+          thrust::fill(
+            thrust::seq, majors + major_offset, majors + major_offset + local_degree, major);
+          thrust::copy(thrust::seq, indices, indices + local_degree, minors + major_offset);
+          if (property) {
+            auto major_input_property  = thrust::get<0>(*property)[idx];
+            auto minor_output_property = thrust::get<1>(*property);
+            thrust::fill(thrust::seq,
+                         minor_output_property + major_offset,
+                         minor_output_property + major_offset + local_degree,
+                         major_input_property);
+          }
+        }
+      });
+  }
+}
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+void decompress_matrix_partition_to_edgelist(
+  raft::handle_t const& handle,
+  matrix_partition_device_view_t<vertex_t, edge_t, weight_t, multi_gpu> const matrix_partition,
+  vertex_t* edgelist_majors /* [OUT] */,
+  vertex_t* edgelist_minors /* [OUT] */,
+  std::optional<weight_t*> edgelist_weights /* [OUT] */,
+  std::optional<std::vector<vertex_t>> const& segment_offsets)
+{
+  auto number_of_edges = matrix_partition.get_number_of_edges();
+
+  decompress_matrix_partition_to_fill_edgelist_majors(
+    handle, matrix_partition, edgelist_majors, segment_offsets);
+  thrust::copy(handle.get_thrust_policy(),
+               matrix_partition.get_indices(),
+               matrix_partition.get_indices() + number_of_edges,
+               edgelist_minors);
+  if (edgelist_weights) {
+    thrust::copy(handle.get_thrust_policy(),
+                 *(matrix_partition.get_weights()),
+                 *(matrix_partition.get_weights()) + number_of_edges,
+                 (*edgelist_weights));
+  }
+}
+
 template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
 void decompress_matrix_partition_to_edgelist(
   raft::handle_t const& handle,
   matrix_partition_device_view_t<vertex_t, edge_t, weight_t, multi_gpu> const matrix_partition,
+  vertex_t const* majors,
```

Comment:
Where are you using this?

Reply:
I am using it on lines [287](https://github.com/rapidsai/cugraph/pull/2117/files/9a52520b4cdccebac3f86bc7e3bd7033e0ab1022#diff-77d25d8411b3eb10b3aca7d2d829bda4836527f68b940c18ffbd85a8a369f98fR287), [306](https://github.com/rapidsai/cugraph/pull/2117/files/9a52520b4cdccebac3f86bc7e3bd7033e0ab1022#diff-77d25d8411b3eb10b3aca7d2d829bda4836527f68b940c18ffbd85a8a369f98fR306), [325](https://github.com/rapidsai/cugraph/pull/2117/files/9a52520b4cdccebac3f86bc7e3bd7033e0ab1022#diff-77d25d8411b3eb10b3aca7d2d829bda4836527f68b940c18ffbd85a8a369f98fR325) and [363](https://github.com/rapidsai/cugraph/pull/2117/files/9a52520b4cdccebac3f86bc7e3bd7033e0ab1022#diff-77d25d8411b3eb10b3aca7d2d829bda4836527f68b940c18ffbd85a8a369f98fR363).